packages/fiori/src/UploadCollection.js

```diff
@@ -0,0 +1,354 @@
+		 * @slot
+		 * @public
+		 */
+		"default": {
+			propertyName: "items",
+			type: HTMLElement,
+		},
+
+		/**
+		 * Defines the <code>ui5-upload-collection</code> header.
+		 *
+		 * @type {HTMLElement[]}
+		 * @slot
+		 * @public
+		 */
+		header: {
+			type: HTMLElement,
+		},
+	},
+	events: /** @lends sap.ui.webcomponents.fiori.UploadCollection.prototype */ {
+		/**
+		 * Fired when the Delete button of any item is pressed.
+		 * <br><br>
+		 * <b>Note:</b> A Delete button is displayed on each item,
+		 * when the <code>ui5-upload-collection</code> <code>mode</code> property is set to <code>Delete</code>.
+		 * @event
+		 * @param {HTMLElement} item The <code>ui5-upload-collection-item</code> which was renamed.
+		 * @public
+		 */
+		fileDeleted: {
+			detail: {
+				item: { type: HTMLElement },
+			},
+		},
+
+		/**
+		 * Fired when selection is changed by user interaction
+		 * in <code>SingleSelect</code> and <code>MultiSelect</code> modes.
+		 *
+		 * @event
+		 * @param {Array} selectedItems An array of the selected items.
+		 * @public
+		 */
+		selectionChange: {
+			detail: {
+				selectedItems: { type: Array },
+			},
+		},
+	},
+};
+
+/**
+ * @class
+ *
+ * <h3 class="comment-api-title">Overview</h3>
+ * This component allows you to represent files before uploading them to a server, with the help of <code>ui5-upload-collection-item</code>.
+ * It also allows you to show already uploaded files.
+ *
+ * <h3>ES6 Module Import</h3>
+ * <code>import @ui5/webcomponents-fiori/dist/UploadCollection.js";</code>
+ *
+ * @constructor
+ * @author SAP SE
+ * @alias sap.ui.webcomponents.fiori.UploadCollection
+ * @extends UI5Element
+ * @tagname ui5-upload-collection
+ * @appenddocs UploadCollectionItem
+ * @public
+ */
+class UploadCollection extends UI5Element {
+	static get metadata() {
+		return metadata;
+	}
+
+	static get render() {
+		return litRender;
+	}
+
+	static get styles() {
+		return UploadCollectionCss;
+	}
+
+	static get template() {
+		return UploadCollectionTemplate;
+	}
+
+	static async onDefine() {
+		await Promise.all([
+			Icon.define(),
+			Label.define(),
+			List.define(),
+			Title.define(),
+			fetchI18nBundle("@ui5/webcomponents"),
+		]);
+	}
+
+	constructor() {
+		super();
+
+		this.i18nBundle = getI18nBundle("@ui5/webcomponents");
+		this._listeners = {
+			dragenter: this._ondragenter.bind(this),
+			dragleave: this._ondragleave.bind(this),
+			dragover: this._ondragover.bind(this),
+			drop: this._ondrop.bind(this),
+		};
+	}
+
+	onEnterDOM() {
+		if (this.noDnd) {
+			return;
+		}
+
+		addUploadCollectionInstance(this);
+	}
+
+	onBeforeRendering() {
+		if (this.noDnd) {
+			return;
+		}
+
+		this._removeDragAndDropListeners();
+	}
+
+	onAfterRendering() {
+		if (this.noDnd) {
+			return;
+		}
+
+		this._addDragAndDropListeners();
+	}
+
+	onExitDOM() {
+		if (this.noDnd) {
+			return;
+		}
+
+		removeUploadCollectionInstance(this);
+		this._removeDragAndDropListeners();
+	}
+
+	_addDragAndDropListeners() {
```

Comment:
You should not register event listeners unless absolutely necessary. You should only bind events in the template and let lit handle the lifecycle. If you need to bind the events conditionally, just bind them and return early in the handlers if the condition is not met. Thus the whole onbefore/onafter rendering logic should go away

Reply:
Done. I was able to simplify the logic even further by adding the subscriptions not on the root div, but on the dedicated div for drag and drop overlay